File: genna/smbase/smbaseJava/__smbase/StateMachine.java
package __smbase;

import java.util.Queue;
import java.util.LinkedList;

public abstract class StateMachine implements Runnable{

    protected Object context;
    protected State innermostActive;
    protected State top;
    private EventQueue othersQueue;
    private Queue<Event> completionQueue;
    
    public final static int COMPLETION = -1;
    public final static int COMPLETED = -2;

    protected abstract void init();

    protected StateMachine(Object c){

	Thread t;

	context = c;
	othersQueue = new EventQueue();
        completionQueue = new LinkedList<Event>();
	top = new TopState(this);

	init();

	t = new Thread(this);

	t.start();
    }

    public void postEvent(Event e){
	new PosterThread(e, othersQueue, 0);
    }

    public void setInnermostActive(State s){
	innermostActive = s;
    }

    public State getInnermostActive(){
	return innermostActive;
    }

    public void run(){
	State source, target, lca;
	Event event;
	EventHandle handle;

	completionQueue.offer(new Event(COMPLETION));

	while(true){

            if(completionQueue.isEmpty())
	        event = othersQueue.get();
	    else
                event = completionQueue.poll();

            source = innermostActive;
	    target = null;

	    do{
		handle = source.handleEvent(event);
		
		if(handle != null){
		    target = handle.getTarget();
		    break;
		}
		
		source = source.getDirectAncestor();
	    }while(source != top);
	    
	    if(target == null)
		continue;

	    if(event.getName() == COMPLETED &&
	       source != innermostActive.getDirectAncestor())
		continue;

	    if(event.getName() == COMPLETION &&
	       source != innermostActive)
		continue;

	    lca = leastCommonAncestor(innermostActive, target);
	    makeExit(innermostActive, lca);
	    executeEffect(handle.getTransitionSequence());
	    innermostActive = target;
	    makeEntry(lca, innermostActive);
	    innermostActive.doActivity();

	}
    }

    protected void postCompletionEvent(Event e){
        completionQueue.offer(e);
    }

    protected void postTimeEvent(Event e, int when){
        new PosterThread(e, othersQueue, when);
    }

    private void makeExit(State from, State to){
	
	State actual = from;

	while(actual != to){
	    actual.exit();
	    actual = actual.getDirectAncestor();
	}
    }

    private void makeEntry(State from, State to){
	
	if(from == to)
	    return;

	makeEntry(from, to.getDirectAncestor());
	to.entry();
    }

    private State leastCommonAncestor(State a, State b){
	
	LinkedList<State> aList, bList, xList;
	State x;

	aList = new LinkedList<State>();
	bList = new LinkedList<State>();

	for(int i = 0 ; i < 2 ; i++){
	    x = i == 0 ? a : b;
	    xList = i == 0 ? aList : bList;

	    while(x != top){
		xList.offer(x);
		x = x.getDirectAncestor();
	    }
	}

	x = aList.getLast();

	while(aList.removeLast() == bList.removeLast()){
	    if(aList.isEmpty() || bList.isEmpty())
		break;

	    x = aList.getLast();
	}

	return x.getDirectAncestor();
    }

    private void executeEffect(LinkedList<Transition> effectSequence){
	
	Transition t;
	
	for(int i = 0 ; i < effectSequence.size() ; i++){
	    t = effectSequence.get(i);
	    
	    if(t.hasEffect())
		t.executeEffect();
	}
    }

    private class TopState extends State {
	
	public TopState(StateMachine sm){
	    super(sm);
	}

	public EventHandle handleEvent(Event e){
	    return new EventHandle(null, null);
	}

	public void entry(){};
	public void doActivity(){};
	public void exit(){};
    }
}

File: genna/smbase/smbaseCpp/__smbase/StateMachine.cpp
#include "StateMachine.h"
#include "PosterThread.h"
#include "TopState.h"

namespace __smbase{

  StateMachine::StateMachine(void *c){
    
    terminated = false;

    context = c;
    top = new TopState(this);

    start();
  }

  StateMachine::~StateMachine(){
    delete top;
  }

  void StateMachine::postEvent(Event *e){

    if(QThread::currentThread() != this){
      othersQueue.put(e);
      return;
    }

    PosterThread *p = new PosterThread(e, &othersQueue, 0);

    terminatedMutex.lock();
    if(!terminated){
      cleanThreads(posterThreads);
      posterThreads.push_back(p);
    } else {
      p->wait();
      delete p;
    }
    terminatedMutex.unlock();
  }

  void StateMachine::setInnermostActive(State *s){
    innermostActive = s;
  }

  State *StateMachine::getInnermostActive(){
    return innermostActive;
  }

  void StateMachine::run(){
    State *source, *target, *lca;
    Event *event;
    EventHandle *handle;

    init();

    completionQueue.push(new Event(COMPLETION));

    while(true){
      
      if(completionQueue.empty())
	event = othersQueue.get();
      else{
	event = completionQueue.front();
	completionQueue.pop();
      }

      if(event->getName() == DIE){
	terminated = true;
	delete event;
	break;
      }

      source = innermostActive;

      do{
	handle = source->handleEvent(event);

	if(handle != 0){
	  target = handle->getTarget();
	  break;
	}

	source = source->getDirectAncestor();
      }while(source != top);

      if(source == top || (event->getName() == COMPLETION &&
	 source != innermostActive) || (event->getName() == COMPLETED &&
	 source != innermostActive->getDirectAncestor())){
	delete event;
        delete handle;
        continue;
      }

      lca = leastCommonAncestor(innermostActive, target);
      makeExit(innermostActive, lca);
      executeEffect(handle->getTransitionSequence());
      innermostActive = target;
      makeEntry(lca, innermostActive);
      innermostActive->doActivity();

      delete event;
      delete handle;
    }

    terminatedMutex.lock();
    waitThreads(posterThreads);
    terminatedMutex.unlock();

    waitThreads(timedThreads);
  }

  void StateMachine::postCompletionEvent(Event *e){
    completionQueue.push(e);
  }

  void StateMachine::postTimeEvent(Event *e, int when){
    cleanThreads(timedThreads);
    timedThreads.push_back(new PosterThread(e, &othersQueue, when));
  }

  void StateMachine::makeExit(State *from, State *to){
    State *actual = from;

    while(actual != to){
      actual->exit();
      actual = actual->getDirectAncestor();
    }
  }

  void StateMachine::makeEntry(State *from, State *to){
    
    if(from == to)
      return;

    makeEntry(from, to->getDirectAncestor());
    to->entry();
  }

  State *StateMachine::leastCommonAncestor(State *a, State *b){
    std::vector<State *> aList, bList, *xList;
    State *x;

    for(int i = 0 ; i < 2 ; i++){
      x = i == 0 ? a : b;
      xList = i == 0 ? &aList : &bList;

      while(x != top){
	xList->push_back(x);
	x = x->getDirectAncestor();
      }
    }

    while((x = aList.back()) == bList.back()){
      aList.pop_back();
      bList.pop_back();
      
      if(aList.empty() || bList.empty())
	break;
    }

    return x->getDirectAncestor();
  }

  void StateMachine::executeEffect(std::vector<Transition *> effectSequence){
    Transition *t;

    for(unsigned int i = 0 ; i < effectSequence.size() ; i++){
      t = effectSequence[i];

      if(t->hasEffect())
	t->executeEffect();
    }
  }

  void StateMachine::cleanThreads(std::vector<PosterThread *> &pt){
    std::vector<PosterThread *>::iterator i = pt.begin();

    while(i != pt.end()){
      if((*i)->isFinished()){
	delete *i;
	*i = 0;
	i = pt.erase(i);
      }
      else{
	i++;
      }
    }

  }

  void StateMachine::waitThreads(std::vector<PosterThread *> pt){

    std::vector<PosterThread *>::iterator i = pt.begin();
    
    while(i != pt.end()){
      if((*i)->isRunning()){
	(*i)->terminate();
	(*i)->wait(10);
      }
      (*i)->wait();
      delete *i;

      i++;
    }
  }
}
